### PR TITLE
oauth2: public keys are served from endpoint

### LINKED ISSUE #38: oauth2: serve public certs
List of certificates which are used for issued  ID Tokens by the oauth2-server library should be served publicly via HTTP endpoint to provide the clients with ability to verify that issued tokens are coming from the correct origin: 

**GET /o/oauth2/v1/certs**
```json
{
 "7c83f5cc1536003737c3759b2b90b1a5d1d4ac65": "-----BEGIN CERTIFICATE-----\n
MIIDJjCCAg6gAwIBAgIIH2ldw266/tIwDQYJKoZIhvcNAQEFBQAwNjE0MDIGA1UE\n
AxMrZmVkZXJhdGVkLXNpZ25vbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTAe\n
Fw0xNzAzMzAxMTQzMzRaFw0xNzA0MDIxMjEzMzRaMDYxNDAyBgNVBAMTK2ZlZGVy\n
YXRlZC1zaWdub24uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wggEiMA0GCSqG\n
SIb3DQEBAQUAA4IBDwAwggEKAoIBAQDG+oh/FeyVeMgk3fd7G2HQPuvg3uNMXjlO\n
0iVBr9dV86wK7DzZylLkLtC63mjJbrKjlTAgMt8VyWpolH0xkaami0WrXjwcz4L+\n
gdOWIcELdpI+PtNFOviVEOqCMuFmpuYQzcc+XgpZdntNUKiaEWJeC0iaS+YwaQTe\n
hRHDV8UWe4igbTlspUSf1Q7ldiap6LWpazC3A0GvU8Ca9LwdIi+AhmkFZKgV6ZDO\n
zDfUwOfDn9C6Tu6KW84ylqPIEsAFRO7RexAtjmATpl0aIssBS5BLm5657B4poIIG\n
rmU0HbYEzk9/Ih1BIstGtXkiR8vYlbFUKvJmegQvaUQN4KgRX5YFAgMBAAGjODA2\n
MAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMBYGA1UdJQEB/wQMMAoGCCsG\n
AQUFBwMCMA0GCSqGSIb3DQEBBQUAA4IBAQC0iJnHCEACPYu19qPvsLar1EkKV0Z2\n
wQDdU2LkdhKgjq+pR8lfjvg5Oh6rW43I4iTyl4Rr4QAQxv/vIE8ucA5dye5inXnK\n
wFCOyOn6jpEyTbQgW5hEqHTT8X5b9DEZna6Ld9SVQrIO4E9bwQbS4LNsD1ejAXme\n
vvCHAUbxm56lnN7NkHz5GjnqArToes1d3aizZ2M93307fuE6ywpebcQ1CgAYih6l\n
8nywdA2wUzN8B4w7KPzk8zi3w64G8pbrvAoXJzLPINfidVuEMkuP/efp/aOW4mSE\n
qd3T4VEP74NEPLMKX5EMCWCg/ixlFVWtoJ6scAFR5sHfqvqewsC35NE7\n
-----END CERTIFICATE-----\n",
 "90a9960da247d3eb884c1d00a3a0030a1282d657": "-----BEGIN CERTIFICATE-----\n
MIIDJjCCAg6gAwIBAgIIH2ldw266/tIwDQYJKoZIhvcNAQEFBQAwNjE0MDIGA1UE\n
AxMrZmVkZXJhdGVkLXNpZ25vbi5zeXN0ZW0uZ3NlcnZpY2VhY2NvdW50LmNvbTAe\n
Fw0xNzAzMzAxMTQzMzRaFw0xNzA0MDIxMjEzMzRaMDYxNDAyBgNVBAMTK2ZlZGVy\n
YXRlZC1zaWdub24uc3lzdGVtLmdzZXJ2aWNlYWNjb3VudC5jb20wggEiMA0GCSqG\n
SIb3DQEBAQUAA4IBDwAwggEKAoIBAQDG+oh/FeyVeMgk3fd7G2HQPuvg3uNMXjlO\n
0iVBr9dV86wK7DzZylLkLtC63mjJbrKjlTAgMt8VyWpolH0xkaami0WrXjwcz4L+\n
gdOWIcELdpI+PtNFOviVEOqCMuFmpuYQzcc+XgpZdntNUKiaEWJeC0iaS+YwaQTe\n
hRHDV8UWe4igbTlspUSf1Q7ldiap6LWpazC3A0GvU8Ca9LwdIi+AhmkFZKgV6ZDO\n
zDfUwOfDn9C6Tu6KW84ylqPIEsAFRO7RexAtjmATpl0aIssBS5BLm5657B4poIIG\n
rmU0HbYEzk9/Ih1BIstGtXkiR8vYlbFUKvJmegQvaUQN4KgRX5YFAgMBAAGjODA2\n
MAwGA1UdEwEB/wQCMAAwDgYDVR0PAQH/BAQDAgeAMBYGA1UdJQEB/wQMMAoGCCsG\n
AQUFBwMCMA0GCSqGSIb3DQEBBQUAA4IBAQC0iJnHCEACPYu19qPvsLar1EkKV0Z2\n
wQDdU2LkdhKgjq+pR8lfjvg5Oh6rW43I4iTyl4Rr4QAQxv/vIE8ucA5dye5inXnK\n
wFCOyOn6jpEyTbQgW5hEqHTT8X5b9DEZna6Ld9SVQrIO4E9bwQbS4LNsD1ejAXme\n
vvCHAUbxm56lnN7NkHz5GjnqArToes1d3aizZ2M93307fuE6ywpebcQ1CgAYih6l\n
8nywdA2wUzN8B4w7KPzk8zi3w64G8pbrvAoXJzLPINfidVuEMkuP/efp/aOW4mSE\n
qd3T4VEP74NEPLMKX5EMCWCg/ixlFVWtoJ6scAFR5sHfqvqewsC35NE7\n
-----END CERTIFICATE-----\n"
}
```

### PR DESCRIPTION
**Pull request by qnislavnachev**
public key now are served publicly by given HTTP endpoint, they
are return as list

Fixed #38

### CODE REVIEW

File: oauth2-server/src/test/java/com/clouway/oauth2/RetrievePublicKeysTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    assertThat(body, is("{\"publicKeys\":[\"cert1\",\"cert2\",\"cert3\"]}"));
+  }
+
+  @Test
+  public void publicKeysIsE() throws Exception {
+    PublicCertsController controller = new PublicCertsController(new FakePublicKeys(new HashMap<String, String>()));
+
+    Response response = controller.ack(request);
+    String body = CharStreams.toString(new InputStreamReader(response.body(), Charsets.UTF_8));
```
Comment (mgenov) [inline oauth2-server/src/test/java/com/clouway/oauth2/RetrievePublicKeysTest.java]: There is a RsPrint class about this.

File: oauth2-server/src/main/java/com/clouway/oauth2/PublicCertsController.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Override
+  public Response ack(Request request) throws IOException {
+    List<String> keyList = keys.getKeys();
+    if (keyList.isEmpty()) {
+      return new RsEmpty();
+    }
+
+    JsonObject o = new JsonObject();
+    o.add("publicKeys", getAsJsonArray(keyList));
```
Comment (mgenov) [inline oauth2-server/src/main/java/com/clouway/oauth2/PublicCertsController.java]: The generated JSON message doesn't look like the prosed in the issue. 


```javascript
{
   "keyid1": "keybody1", 
   "keyid2": "keybody2"
}
```

**Review by mgenov — CHANGES_REQUESTED** [review]

File: oauth2-server/src/test/java/com/clouway/oauth2/RetrievePublicKeysTest.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    PublicCertsController controller = new PublicCertsController(new FakePublicKeys(keys));
+
+    Response response = controller.ack(request);
+    String body = CharStreams.toString(new InputStreamReader(response.body(), Charsets.UTF_8));
+
+    assertThat(body, is("{\"publicKeys\":[\"cert1\",\"cert2\",\"cert3\"]}"));
+  }
+
+  @Test
+  public void publicKeysIsE() throws Exception {
```
Comment (mgenov) [inline oauth2-server/src/test/java/com/clouway/oauth2/RetrievePublicKeysTest.java]: isE?

**Comment by qnislavnachev** [comment]
I am sorry PR is not ready for review yet.. my bad

**Comment by qnislavnachev** [comment]
PR is ready for review. Comments was fixed.

**Comment by qnislavnachev** [comment]
There is some issue with the continuous integration server that I will check

File: oauth2-server/src/main/java/com/clouway/oauth2/PublicKeys.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.google.common.base.Optional;
+
+import java.security.Key;
+import java.util.Map;
+
+/**
+ * @author Ianislav Nachev <qnislav.nachev@clouway.com>
+ */
+public interface PublicKeys {
```
Comment (mgenov) [inline oauth2-server/src/main/java/com/clouway/oauth2/PublicKeys.java]: Add comments to this public interface, as it's going to be used by client code which need to know what's the purpose of it.

File: oauth2-server/src/test/java/com/clouway/oauth2/RetrievePublicKeysTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * @author Ianislav Nachev <qnislav.nachev@clouway.com>
+ */
+public class RetrievePublicKeysTest {
+
+  Request request;
+
+  @Test
+  public void retrievePublicKeys() throws Exception {
+    Map<String, String> keys = new HashMap<String, String>() {{
```
Comment (mgenov) [inline oauth2-server/src/test/java/com/clouway/oauth2/RetrievePublicKeysTest.java]: Use sorted map like TreeMap to ensure that order is preserved.

**Review by mgenov — CHANGES_REQUESTED** [review]

**Comment by qnislavnachev** [comment]
PR notes fixed need another review.

**Review by mgenov — APPROVED** [review]
LGTM !

**Pull request merged**